import os


### snakemake_workflows initialization ########################################
maindir = os.path.dirname(os.path.dirname(workflow.basedir))
sys.path.append(os.path.join(maindir, "shared"))

import common_functions as cf

# load tool paths
globals().update(cf.load_paths(os.path.join(maindir, "shared", "paths.yaml"),maindir,config["verbose"]))
# load config file
globals().update(cf.load_configfile(workflow.overwrite_configfile,config["verbose"]))
# load organism-specific data, i.e. genome indices, annotation, etc.
globals().update(cf.load_organism_data(genome,maindir,config["verbose"]))

# do workflow specific stuff now
include: os.path.join(workflow.basedir, "internals.snakefile")

### include modules of other snakefiles ########################################
################################################################################
# deepTools ChIP
# include: os.path.join(maindir, "shared", "rules", "deepTools_ChIP.snakefile")

# ATAC-seq open chromatin
include: os.path.join(maindir, "shared", "rules", "ATAC.snakefile")


### execute before workflow starts #############################################
################################################################################
onstart:
    if "verbose" in config and config["verbose"]:
        print("--- Workflow parameters --------------------------------------------------------")
        print("samples:", samples)
        print("paired:", paired)
        print("ATAC fragment cutoff: ", atac_fragment_cutoff)
        print("-" * 80, "\n")

        print("--- Environment ----------------------------------------------------------------")
        print("$TMPDIR: ",os.getenv('TMPDIR', ""))
        print("$HOSTNAME: ",os.getenv('HOSTNAME', ""))
        print("-" * 80, "\n")

        print("--- Genome ---------------------------------------------------------------------")
        print("Genome:", genome)
        print("Effective genome size:", genome_size)
        print("Genome FASTA:", genome_fasta)
        print("Genome index:", genome_index)
        print("Genome 2bit:", genome_2bit)
        print("Bowtie2 index:", bowtie2_index)
        print("Gene annotation BED:", genes_bed)
        print("Gene annotation GTF:", genes_gtf)
        print("Blacklist regions BED:", blacklist_bed)
        print("-" * 80, "\n")

### main rule ##################################################################
################################################################################
rule all:
    input:
        # #
        # # ## prepare for ATAC peak calling
        # # # 1) sort by query
        # # # 2) convert bam to bedpe
        # # # 3) filter fragments
        expand("peaks_openChromatin/{sample}.filtered.sorted.bam", sample = samples),
        expand("peaks_openChromatin/{sample}.all.bedpe", sample = samples),
        expand("peaks_openChromatin/{sample}.openchrom.bedpe", sample = samples),
        expand("peaks_openChromatin/{sample}.all.fragdistr", sample=samples),
        # #
        # # ## perform peak calling
        expand("peaks_openChromatin/openchromatin_{sample}_peaks.narrowPeak", sample = samples),
        expand('peaks_openChromatin/openchromatin_{sample}_treat_pileup.bdg', sample = samples),
        expand('peaks_openChromatin/openchromatin_{sample}_control_lambda.bdg', sample = samples),
        expand("peaks_openChromatin/openchromatin_{sample}_{condition}.bw", sample = samples, condition = ('control_lambda','treat_pileup'))



### execute after workflow finished ############################################
################################################################################
onsuccess:
    if "verbose" in config and config["verbose"]:
        print("\n--- ATAC OpenChrom workflow finished successfully! --------------------------------\n")
