

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="python" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="python" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Change Log &mdash; snakePipes 0.6.2 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="HiC" href="content/workflows/HiC.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> snakePipes
          

          
          </a>

          
            
            
              <div class="version">
                0.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="content/setting_up.html">Setting up the workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/DNA-mapping.html">DNA-mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/ChIP-seq.html">ChIP-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/RNA-seq.html">RNA-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/scRNA-seq.html">scRNA-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/HiC.html">HiC</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Change Log</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">snakePipes</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Change Log</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/ChangeLog.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="change-log">
<h1>Change Log<a class="headerlink" href="#change-log" title="Permalink to this headline">Â¶</a></h1>
<p>The history of development of the workflows is listed below, along with github IDs of
people with most (but not all) contribution to the changes.</p>
<p>version 0.0.1 - March 23, 2016 - &#64;kilpert
- initial version</p>
<p>version 0.1.0 - June 15, 2016 - &#64;asrichter
- added --fastqc and --bw-binsize parameters to DNA-mapping wrapper script
- additional organisms are now supported by adding new genome.py files
- defined (effective) genome size as (genome length)-(number of 'N's) in genome.py files
- simplified cluster submission by taking 'threads' parameter from rule definition, removed cluster.yaml
- added DNA-mapping example configuration yaml file
- added function get_fragment_length() to DNA-mapping internals.snakefile to parse median insert size from Picard output
- added deepTools_qc.snakefile as one common snakefile for all deepTools rules,</p>
<blockquote>
<div>moved all deepTools rules from ChIP-seq Snakefile and computeGCBias.snakefile</div></blockquote>
<ul class="simple">
<li>moved include statements of module snakefiles from rules.snakefile to DNA-mapping Snakefile to simplify structure, removed rules.snakefile</li>
<li>renamed functions do_TrimGalore(trim) and do_InsertSizeMetrics(paired) in DNA-mapping Snakefile</li>
<li>renamed load_organisms.snakefile to load_organism_data.py</li>
<li>removed debugging.snakefile as it was not used by any pipeline</li>
<li>removed --local-cores parameter from all wrapper scripts as there are no local snakemake rules defined</li>
<li>many small other changes</li>
</ul>
<p>version 0.1.1 - June 17, 2016 -  &#64;asrichter
- added option to run workflow locally instead of cluster submission
- FASTQ.snakefile replaces FASTQ_symlink.snakefile and FASTQ_downsample.snakefile
- several small changes</p>
<p>version 0.2.0 - June 22, 2016 -  &#64;asrichter
- added filtering option to filter BAM files for duplication, proper pairs and MAPQ
- added variable 'outdir' to configuration
- many small changes</p>
<p>version 0.3.0 - June 24, 2016 -  &#64;asrichter
- rewrote ChIP-seq workflow completely including wrapper script ChIP-seq
- added histoneHMM for calling broadly enriched regions
- added MACS2 peak quality controls
- revised example.config.yaml files
- moved function get_fragment_length() to common_functions.py
- positional instead of required optional command line arguments
- many small changes</p>
<p>version 0.3.1 - June 25, 2016 -  &#64;asrichter
- run Picard quality control on unfiltered BAM files
- added --gcbias parameter to DNA-mapping wrapper script to run computeGCBias optionally
- replaced --input-dir and --output-dir by --working-dir parameter in ChIP-seq</p>
<blockquote>
<div>wrapper script to specify the working directory, which is output directory of
the pipeline and must also contain the DNA-mapping pipeline output files</div></blockquote>
<ul class="simple">
<li>bugfixes</li>
</ul>
<p>version 0.3.2 - June 27, 2016 -  &#64;asrichter
- added generation of QC reports for all samples to ChIP-seq pipeline
- added consistency check for ChIP-seq pipeline whether all required input files exist for all samples
- added peak count to MACS2 peak quality controls</p>
<p>version 0.3.2.1 - June 27, 2016 - Andreas Richter
- added documentation to README.md
- moved R library</p>
<p>version 0.4 - 2016 - &#64;kilpert, &#64;steffenheyne</p>
<p>version 0.5 - 2017 - &#64;steffenheyne, &#64;kilpert, &#64;mirax87
- major cleanup and refactoring of wrappers and code structure (but not rules)
- scRNAseq workflow added
- using yaml config files all over, ie.</p>
<blockquote>
<div><ol class="arabic simple">
<li>all wrappers write out a config yaml</li>
<li>Snakefile configuration only depends on provided '--configfile'</li>
</ol>
</div></blockquote>
<ul class="simple">
<li>better handling and and more usage of common_functions</li>
<li>all genome config files converted to yaml, function to load genome information</li>
<li>own genome file can be provided instead of only pre-configured ones</li>
<li>paths config file converted to yaml</li>
<li>eliminate organism specific t2g and symbol files, they are now created by rule from gtf</li>
<li>check that sleuth is only run if replicates are available</li>
<li>bugfix: only provided wrapper arguments overwrite either defaults or configfile options, before
a not provided wrapper option (but in config file provided) would be overwritten by default,
now there is true hierachy: defaults-&gt;configfile-&gt;wrapper !</li>
</ul>
<p>version 0.6 (a.k.a Tiger RattleSnake) - Sept 2017 - &#64;vivekbhr
- MAJOR CHANGES:</p>
<blockquote>
<div><ul class="simple">
<li>Allele-Specific mapping : Allele-specific DNA and RNA-mapping is now possible and both ChIP-Seq and RNA-seq pipeline can handle &quot;allele_mapping&quot; mode.</li>
<li>Differential binding : Differential binding can be performed using CSAW, both normal and allele-specific.</li>
<li>MultiQC : MultiQC summarizes QC results for DNA-mapping and RNA-mapping workflows.</li>
<li>R package : New R package called snakediff contains all functions for differential binding and expression. The DE functions have been
transformed into clean and consistent layout, with docs.</li>
<li>Deeptools commands : All deeptools commands now moved into common deeptools_cmds module, and are shared between functions.</li>
</ul>
</div></blockquote>
<ul>
<li><p class="first">MINOR CHANGES:
- Wrappers have been formatted for easier readability.
- chr X, Y, M and unmapped scaffolds are now ignored for normalization in bamCoverage (DNA-mapping) and bamCompare (Chip-Seq).
- bamCompare outputs are named with suffix &lt;control&gt;, instead of a static suffix &quot;Input&quot;, this allows the workflow to run with different controls</p>
<blockquote>
<div><p>eg. Input, H3 and keep results in same dir. However, MACS2 would still over-write the output files from previous run (needs fix).</p>
</div></blockquote>
</li>
</ul>
<p>Version 0.6.1</p>
<ul class="simple">
<li>MINOR CHANGES:
- The allele-specific option is no longer on by default (it was &#64;vivekbhr's fault)</li>
</ul>
<p>Version 0.6.2</p>
<ul class="simple">
<li>MINOR CHANGES:
- Explicitly define which snakemake version to use</li>
</ul>
<p>Version 0.7 (a.k.a Green Mamba) - Nov 2017 - &#64;vivekbhr</p>
<ul class="simple">
<li>MAJOR CHANGES:
- Read the Docs integration
- New workflow Hi-C, from mapping to TAD calling, using BWA and HiCExplorer</li>
</ul>
<p>Version 0.7.1 - Jan. 12 2018</p>
<ul class="simple">
<li>MINOR CHANGES:
- Fixed issue #57 again via pull request #79, which was related to single-end ChIPseq processing</li>
</ul>
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference external" href="https://github.com/maxplanck-ie/snakemake_workflows">code &#64; github</a>.</td>
</tr>
</tbody>
</table>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="content/workflows/HiC.html" class="btn btn-neutral" title="HiC" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, MPI-IE.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.6.2',
            LANGUAGE:'python',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>