

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="python" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="python" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>snakePipes &mdash; snakePipes 0.6.2 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Setting up the workflows" href="content/setting_up.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="#" class="icon icon-home"> snakePipes
          

          
          </a>

          
            
            
              <div class="version">
                0.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="content/setting_up.html">Setting up the workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/DNA-mapping.html">DNA-mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/ChIP-seq.html">ChIP-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/RNA-seq.html">RNA-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/scRNA-seq.html">scRNA-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/HiC.html">HiC</a></li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/WGBS.html">WGBS</a></li>
<li class="toctree-l1"><a class="reference internal" href="ChangeLog.html">Change Log</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">snakePipes</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>snakePipes</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="snakepipes">
<h1>snakePipes<a class="headerlink" href="#snakepipes" title="Permalink to this headline">¶</a></h1>
<p>snakePipes are pipelines built using snakemake for the analysis of various sequencing datasets.</p>
<div class="section" id="below-is-the-list-of-pipelines-available-in-snakepipes">
<h2>Below is the list of pipelines available in snakePipes<a class="headerlink" href="#below-is-the-list-of-pipelines-available-in-snakepipes" title="Permalink to this headline">¶</a></h2>
<table border="1" class="docutils">
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Pipeline</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><a class="reference internal" href="content/workflows/DNA-mapping.html#dna-mapping"><span class="std std-ref">DNA-mapping</span></a></td>
<td>Basic DNA mapping using bowtie2, filter mapped files, QC and create coverage plots</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="content/workflows/ChIP-seq.html#chip-seq"><span class="std std-ref">ChIP-seq</span></a></td>
<td>Use the DNA mapping output and run ChIP/Input normalization and peak calling</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="content/workflows/ATAC-seq.html#atac-seq"><span class="std std-ref">ATAC-seq</span></a></td>
<td>Use the DNA mapping output and detect open chromatin regions from ATAC-seq</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="content/workflows/RNA-seq.html#rna-seq"><span class="std std-ref">RNA-seq</span></a></td>
<td>RNA-Seq workflow : From mapping to differential expression using DEseq2</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="content/workflows/scRNA-seq.html#scrna-seq"><span class="std std-ref">scRNA-seq</span></a></td>
<td>Single-cell RNA-Seq workflow : From mapping to differential expression</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="content/workflows/HiC.html#hic"><span class="std std-ref">HiC</span></a></td>
<td>Hi-C analysis workflow, from mapping to TAD calling</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="content/workflows/WGBS.html#wgbs"><span class="std std-ref">WGBS</span></a></td>
<td>WGBS analysis workflow, from mapping to DMR calling and differential methylation analysis</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="quick-start">
<h2>Quick start<a class="headerlink" href="#quick-start" title="Permalink to this headline">¶</a></h2>
<p>The desing of all workflows, as well as configuration is the same. Here's an example with ChIP-seq data.</p>
<p>A <strong>typical ChIP-seq analysis of human samples</strong> starting from paired-end FASTQ files in the directory <cite>input-dir</cite>:</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ ls /path/to/input-dir/
my_H3K27ac_sample_R1.fastq.gz  my_H3K27me3_sample_R1.fastq.gz  my_Input_sample_R1.fastq.gz
my_H3K27ac_sample_R2.fastq.gz  my_H3K27me3_sample_R2.fastq.gz  my_Input_sample_R2.fastq.gz
$
$ snakemake_workflows/DNA-mapping \
      -i /path/to/input-dir -o /path/to/output-dir --dedup hs37d5 &amp;&amp; \
  snakemake_workflows/ChIP-seq chip-seq.config.yaml \
      -d /path/to/outputdir hs37d5
</pre></div>
</div>
<p>Here, hs37d5 is the name of the genome. The yaml file corresponding to this genome should exist under <cite>/shared/organisms/hs37d5.yaml</cite>.
This yaml file should have paths to the required genome fasta, index, GTF and other annotations (see <strong>Genome configuration file</strong> below).</p>
<p>All individual jobs of the workflow will be submitted to the Grid engine using the command specified under /shared/cluster.yaml.
To run the workflow locally, use the parameter <cite>--local</cite> for local mode and the parameter <cite>-j 48</cite> to specify the maximal
number of used CPU threads (here: 48) or concurrent running Slurm jobs (actual used threads are defined in each rule).</p>
<p>A <strong>configuration file is required for the ChIP-seq workflow</strong> and should adhere to the following style :
<strong>IMPORTANT: Use only whitespace, but NO TAB indentation in this file:</strong></p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat chip-seq.config.yaml
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">################################################################################</span>
<span class="c1"># Please specify all ChIP samples plus their matching control/chromatin input</span>
<span class="c1"># sample.</span>
<span class="c1"># Specify for each ChIP sample whether the IP target results in broad/mixed-type</span>
<span class="c1"># enrichment (most histone marks, e.g. H3K4me1, H3K36me3, H3K9me3, H3K27me3)</span>
<span class="c1"># or not. In the latter case, the enrichment is instead punctuate/narrow</span>
<span class="c1"># (e.g. TFs, active histone marks as H3K27ac or H3K4me3).</span>
<span class="c1">#</span>
<span class="c1"># IMPORTANT: Use only whitespace, but NO TAB indentation in this YAML file!</span>
<span class="c1">################################################################################</span>
<span class="n">chip_dict</span><span class="p">:</span>
  <span class="n">my_H3K27ac_sample</span><span class="p">:</span>
    <span class="n">control</span><span class="p">:</span> <span class="n">my_Input_sample</span>
    <span class="n">broad</span><span class="p">:</span> <span class="kc">False</span>
  <span class="n">my_H3K27me3_sample</span><span class="p">:</span>
    <span class="n">control</span><span class="p">:</span> <span class="n">my_Input_sample</span>
    <span class="n">broad</span><span class="p">:</span> <span class="kc">True</span>
</pre></div>
</div>
</div>
<div class="section" id="genome-configuration-file">
<h2>Genome configuration file<a class="headerlink" href="#genome-configuration-file" title="Permalink to this headline">¶</a></h2>
<p>Any organism can be supported in the workflows by adding a genome configuration file <cite>my_organism.yaml</cite> in the
following style to the <cite>snakemake_workflows</cite> directory:</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat snakemake_workflows/shared/organisms/hs37d5.yaml
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">genome_size</span><span class="p">:</span> <span class="mi">2900338458</span>
<span class="n">genome_fasta</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/genome_fasta/genome.fa&quot;</span>
<span class="n">genome_index</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/genome_fasta/genome.fa.fai&quot;</span>
<span class="n">genome_2bit</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/genome_fasta/genome.2bit&quot;</span>
<span class="n">bowtie2_index</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/BowtieIndex/genome&quot;</span>
<span class="n">hisat2_index</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/HISAT2Index/genome&quot;</span>
<span class="n">known_splicesites</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/gencode/release_19/HISAT2/splice_sites.txt&quot;</span>
<span class="n">star_index</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/STARIndex/&quot;</span>
<span class="n">genes_bed</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/gencode/release_19/genes.bed&quot;</span>
<span class="n">genes_gtf</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/gencode/release_19/genes.gtf&quot;</span>
<span class="n">blacklist_bed</span><span class="p">:</span> <span class="s2">&quot;/SOMEPATH/hs37d5_ensembl/ENCODE/hs37d5_extended_Encode-blacklist.bed&quot;</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If no blacklist regions are available for your organism of interest, leave <cite>blacklist_bed:</cite> empty</p>
</div>
</div>
<div class="section" id="contents">
<h2>Contents:<a class="headerlink" href="#contents" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="content/setting_up.html">Setting up the workflows</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/setting_up.html#set-up-slurm-and-snakemake">Set up slurm and snakemake</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/setting_up.html#edit-the-paths-to-the-required-programs">Edit the paths to the required programs</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/setting_up.html#configure-the-organisms">Configure the organisms</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/DNA-mapping.html">DNA-mapping</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/DNA-mapping.html#configuration-file">Configuration file</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/DNA-mapping.html#structure-of-output-directory">Structure of output directory</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/ChIP-seq.html">ChIP-seq</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/ChIP-seq.html#configuration-file">Configuration file</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/ChIP-seq.html#input-requirements">Input requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/ChIP-seq.html#structure-of-output-directory">Structure of output directory</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/RNA-seq.html">RNA-seq</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/RNA-seq.html#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/RNA-seq.html#Named Arguments">Named Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/RNA-seq.html#required arguments">required arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/RNA-seq.html#general arguments">general arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/RNA-seq.html#Optional arguments">Optional arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/RNA-seq.html#allele-specific mapping arguments">allele-specific mapping arguments</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/scRNA-seq.html">scRNA-seq</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/scRNA-seq.html#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/scRNA-seq.html#required arguments">required arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/scRNA-seq.html#general arguments">general arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/scRNA-seq.html#Optional arguments">Optional arguments</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/HiC.html">HiC</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/HiC.html#input-requirements-and-outputs">Input requirements and outputs:</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/HiC.html#structure-of-output-directory">Structure of output directory</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="content/workflows/WGBS.html">WGBS</a><ul>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/WGBS.html#input-requirements-and-outputs">Input requirements and outputs:</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/WGBS.html#configuration-file">Configuration file</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/WGBS.html#structure-of-output-directory">Structure of output directory</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/WGBS.html#usage-example">Usage example</a></li>
<li class="toctree-l2"><a class="reference internal" href="content/workflows/WGBS.html#argparse">Argparse</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="ChangeLog.html">Change Log</a></li>
</ul>
</div>
</div>
<div class="section" id="citation">
<h2>Citation<a class="headerlink" href="#citation" title="Permalink to this headline">¶</a></h2>
<img alt="_images/logo_mpi-ie.jpg" src="_images/logo_mpi-ie.jpg" />
<p>This tool suite is developed by the <a class="reference external" href="http://www.ie-freiburg.mpg.de/bioinformaticsfac">Bioinformatics Unit</a>
at the <a class="reference external" href="http://www.ie-freiburg.mpg.de/">Max Planck Institute for Immunobiology and Epigenetics</a>, Freiburg.</p>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference external" href="https://github.com/maxplanck-ie/snakemake_workflows">code &#64; github</a>.</td>
</tr>
</tbody>
</table>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="content/setting_up.html" class="btn btn-neutral float-right" title="Setting up the workflows" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, MPI-IE.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.6.2',
            LANGUAGE:'python',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>