---
author: "`r Sys.info()[length(Sys.info())-1]`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
    html_document:
        fig_caption: yes
params:
    outdir: "Stats_reports"
    output_file: "Stats_reports/Stats_report.html"
    sample_sheet: "dummy_sample_sheet"
    fdr: 0.05
    lfc: 1
    pipeline: ChIP-seq
title: "`r paste0('CSAW differential binding stats report')`"
---

```{r setup, include=FALSE}
.libPaths(R.home("library"))

knitr::opts_chunk$set(echo = FALSE, out.width = '30%',out.height='30%')
source(params$input_func)
```


### Overview

This report summarizes statistical analysis of binding events using group information provided in sample sheet `r params$sample_sheet`. 

### Analysis input

```{r}
DBdata<-paste0("CSAW_",gsub("\\.[a-z]{3}$","",params$sample_sheet))
if(is.na(file.info(DBdata)$size))|file.info(DBdata)$size!=0){
    message("Output of differential binding analysis not found or empty!")}else{
    load(DBdata)}
```

There were `r length(chip_object$windowCounts)` genomic bins of size `r width(chip_object$windowCounts)[1]` intersecting MACS2 peaks.   
Size factors were calculated via TMM normalization of read counts obtained for 10kb genomic bins and amounted to `r chip_object$normFactors`.   

### Statistical analysis

Statistical model was fit according to the following design:

```{r}
chip_object$design
```

The original windows were merged into `r length(chip_results$mergedRegions)` regions, for which p values were combined and FDR estimates were obtained after adjustment for multiple testing. 
The maximal absolute log fold change of a window belonging to a merged region was reported as best log fold change for that region.

### Filtering of results

```{r}
merged<-chip_results$mergedRegions
tabcom<-chip_results$combinedPvalues
merged$region$score <- -10*log10(tabcom$FDR)
names(merged$region) <- paste0("region", 1:length(merged$region))
tabcom$name <- names(merged$region)
tabx<-as.data.frame(merged$region,stringsAsFactors=FALSE)
tabx$name<-rownames(tabx)
full_res<-as.data.frame(merge(x=tabx,y=tabcom,by.x="name",by.y="name"),stringsAsFactors=FALSE)
fdrcutoff<-as.numeric(params$fdr)
lfccutoff<-as.numeric(params$lfc)
full_res.filt<-subset(full_res,(FDR<=fdrcutoff)&(abs(best.logFC)>=lfccutoff))
```

Applying FDR threshold of 'r params$fdr' and absolute log fold change threshold of `r params$lfc` result in a final list of `r nrow(full_res.filt)` merged regions.   
Counts of regions showing direction change were as follows:

```{r}
table(full_res.filt$direction)
```

### Visualizing the results

Deeptools matrices were computed for the filtered lists of merged regions showing different change directions and visualized as heatmaps with summary profiles.

Heatmaps obtained using log2 ratio of ChIP signal over input were as follows:

```{r, caption=c("UP","DOWN","MIXED")}
if(!params$pipeline %in% "chip-seq"){message("ChIP signal enrichment over input is only available for ChIP-seq workflow!")}else{
    knitr::includeGraphics(file.path(outdir,"CSAW.UP.log2r.heatmap.png"))
    knitr::includeGraphics(file.path(outdir,"CSAW.DOWN.log2r.heatmap.png"))
    knitr::includeGraphics(file.path(outdir,"CSAW.MIXED.log2r.heatmap.png"))}
```
Heatmaps obtained using depth-normalized coverage were as follows:

```{r, caption=c("UP","DOWN","MIXED")}
knitr::includeGraphics(file.path(outdir,"CSAW.UP.cov.heatmap.png"))
knitr::includeGraphics(file.path(outdir,"CSAW.DOWN.cov.heatmap.png"))
knitr::includeGraphics(file.path(outdir,"CSAW.MIXED.cov.heatmap.png"))
```

Tables with differentially bound regions annotated with distance to nearest gene can be found in folder `r paste0("AnnotatedResults_",gsub("\\.[a-z]{3}$","",params$sample_sheet))`.


  
