

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="python" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="python" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>RNA-seq &mdash; snakePipes 0.6.2 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="scRNA-seq" href="scRNA-seq.html" />
    <link rel="prev" title="ChIP-seq" href="ChIP-seq.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> snakePipes
          

          
          </a>

          
            
            
              <div class="version">
                0.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../setting_up.html">Setting up the workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="DNA-mapping.html">DNA-mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="ChIP-seq.html">ChIP-seq</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">RNA-seq</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#required arguments">required arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#general arguments">general arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Optional arguments">Optional arguments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#allele-specific mapping arguments">allele-specific mapping arguments</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="scRNA-seq.html">scRNA-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="HiC.html">HiC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ChangeLog.html">Change Log</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">snakePipes</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
      <li>RNA-seq</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/content/workflows/RNA-seq.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="rna-seq">
<span id="id1"></span><h1>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">¶</a></h1>
<p><p>RNA-seq workflow v0.6 - MPI-IE workflow for RNA mapping and analysis</p>
<dl class="docutils">
<dt>usage example:</dt>
<dd>RNA-seq -i input-dir -o output-dir mm10</dd>
</dl>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">RNA</span><span class="o">-</span><span class="n">seq</span> <span class="o">-</span><span class="n">i</span> <span class="n">INDIR</span> <span class="o">-</span><span class="n">o</span> <span class="n">OUTDIR</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">CONFIGFILE</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">cluster_configfile</span> <span class="n">CLUSTER_CONFIGFILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">j</span> <span class="n">INT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">local</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">snakemake_options</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">tempdir</span> <span class="n">TEMPDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">DAG</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span> <span class="n">MODE</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">downsample</span> <span class="n">INT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">trim_prg</span> <span class="p">{</span><span class="n">cutadapt</span><span class="p">,</span><span class="n">trimgalore</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">trim</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">trim_options</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">fastqc</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">library_type</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">mapping_prg</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">star_options</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">hisat_options</span> <span class="n">STR</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">salmon_index_options</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">featurecounts_options</span> <span class="n">STR</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">filter_annotation</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">DE</span> <span class="n">SAMPLE_INFO</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">bw</span><span class="o">-</span><span class="n">binsize</span> <span class="n">INT</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">dnaContam</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">plotFormat</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">VCFfile</span> <span class="n">STR</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">strains</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">SNPfile</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">Nmasked_index</span> <span class="n">STR</span><span class="p">]</span>
               <span class="n">GENOME</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h2>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>GENOME</kbd></td>
<td>genome acronym of target organism (supported: 'dm3', 'dm6', 'hs37d5', 'mm9', 'mm10', 'SchizoSPombe_ASM294v2')</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Named Arguments">
<h2>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--plotFormat</kbd></td>
<td><p class="first">Possible choices: png, pdf, None</p>
<p class="last">Format of the output plots from deeptools. Select 'none' for no plot (default: 'None')</p>
</td></tr>
</tbody>
</table>
</div>
<div class="section" id="required arguments">
<h2>required arguments<a class="headerlink" href="#required arguments" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd>-i, --input-dir</kbd></td>
</tr>
<tr><td>&#160;</td><td>input directory containing the FASTQ files, either paired-end OR single-end data</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-o, --output-dir</kbd></td>
</tr>
<tr><td>&#160;</td><td>output directory</td></tr>
</tbody>
</table>
</div>
<div class="section" id="general arguments">
<h2>general arguments<a class="headerlink" href="#general arguments" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-v, --verbose</kbd></td>
<td>verbose output (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-c, --configfile</kbd></td>
</tr>
<tr><td>&#160;</td><td>configuration file: config.yaml. In absence of a config file, the default options from the workflows would be selected from /workflows/&lt;workflow_name&gt;/defaults.yaml (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--cluster_configfile</kbd></td>
</tr>
<tr><td>&#160;</td><td>configuration file for cluster usage. In absence, the default options from shared/cluster.yaml and workflows/[workflow]/cluster.yaml would be selected (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>-j, --jobs</kbd></td>
<td>maximum number of concurrently submitted Slurm jobs / cores if workflow is run locally (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--local</kbd></td>
<td><p class="first">run workflow locally; default: jobs are submitted to Slurm queue (default: 'False')</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--snakemake_options</kbd></td>
</tr>
<tr><td>&#160;</td><td>Snakemake options to be passed directly to snakemake, e.g. use --snakemake_options='--dryrun --rerun-incomplete --unlock --forceall'. (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--tempdir</kbd></td>
<td>used prefix path for temporary directory created via mktemp. Created temp dir gets exported as $TMPDIR and is removed at the end of this wrapper! (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--DAG</kbd></td>
<td><p class="first">If specified, a file called pipeline.pdf is produced in the output directory that shows the rules used and their relationship to each other.</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
<div class="section" id="Optional arguments">
<h2>Optional arguments<a class="headerlink" href="#Optional arguments" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-m, --mode</kbd></td>
<td>workflow running modes (available: 'mapping-free, mapping, allelic-mapping, deepTools_qc') (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--downsample</kbd></td>
<td>downsample the given number of reads randomly from each FASTQ file (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--trim_prg</kbd></td>
<td><p class="first">Possible choices: cutadapt, trimgalore</p>
<p class="last">trimming program: cutadapt or trimgalore (default: 'None')</p>
</td></tr>
<tr><td class="option-group">
<kbd>--trim</kbd></td>
<td>Activate fastq read trimming. If activated, Illumina adaptors are trimmed by default. Additional parameters can be specified under --trim_options (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--trim_options</kbd></td>
<td>Additional option string for trimming program of choice. (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--fastqc</kbd></td>
<td>run FastQC read quality control (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--library_type</kbd></td>
<td>user provided library type strand specificity. featurCounts style: 0, 1, 2 (Illumina TruSeq); default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--mapping_prg</kbd></td>
<td>Program used for mapping: STAR or HISAT2 (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--star_options</kbd></td>
<td>STAR option string, e.g.: '--twopassMode Basic' (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--hisat_options</kbd></td>
</tr>
<tr><td>&#160;</td><td>HISAT2 option string (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--salmon_index_options</kbd></td>
</tr>
<tr><td>&#160;</td><td>Salmon index options, e.g. '--type fmd' (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--featurecounts_options</kbd></td>
</tr>
<tr><td>&#160;</td><td>featureCounts option string. The options '-p -B' are always used for paired-end data (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--filter_annotation</kbd></td>
</tr>
<tr><td>&#160;</td><td>filter annotation GTF by grep for use with Salmon, e.g. use --filter_annotation='-v pseudogene'; default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--DE</kbd></td>
<td>Information on samples (required for DE analysis); see 'snakemake_workflows/shared/tools/sampleInfo.example.tsv' for example. The column names in the tsv files are 'sample' and 'condition'.The first entry defines which group of samples are control.  This way, the order of comparison and likewise the sign of values can be changed. The DE analysis might fail if your sample names begin with a number. So watch out for that! (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--bw-binsize</kbd></td>
<td>bin size of output files in bigWig format (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--dnaContam</kbd></td>
<td>Returns a plot which presents the proportion of the intergenic reads (default: 'None')</td></tr>
</tbody>
</table>
</div>
<div class="section" id="allele-specific mapping arguments">
<h2>allele-specific mapping arguments<a class="headerlink" href="#allele-specific mapping arguments" title="Permalink to this headline">¶</a></h2>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--VCFfile</kbd></td>
<td><p class="first">VCF file to create N-masked genomes (default: 'None')</p>
<p class="last">Default: &quot;&quot;</p>
</td></tr>
<tr><td class="option-group">
<kbd>--strains</kbd></td>
<td><p class="first">Name or ID of SNP strains separated by comma (default: 'None')</p>
<p class="last">Default: &quot;&quot;</p>
</td></tr>
<tr><td class="option-group">
<kbd>--SNPfile</kbd></td>
<td><p class="first">File containing SNP locations (default: 'None')</p>
<p class="last">Default: &quot;&quot;</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--Nmasked_index</kbd></td>
</tr>
<tr><td>&#160;</td><td><p class="first">N-masked index of the reference genome (default: 'None')</p>
<p class="last">Default: &quot;&quot;</p>
</td></tr>
</tbody>
</table>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference external" href="https://github.com/maxplanck-ie/snakemake_workflows">code &#64; github</a>.</td>
</tr>
</tbody>
</table>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="scRNA-seq.html" class="btn btn-neutral float-right" title="scRNA-seq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="ChIP-seq.html" class="btn btn-neutral" title="ChIP-seq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, MPI-IE.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.6.2',
            LANGUAGE:'python',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>