

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="python" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="python" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>WGBS &mdash; snakePipes 0.6.2 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Change Log" href="../../ChangeLog.html" />
    <link rel="prev" title="HiC" href="HiC.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> snakePipes
          

          
          </a>

          
            
            
              <div class="version">
                0.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../setting_up.html">Setting up the workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="DNA-mapping.html">DNA-mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="ChIP-seq.html">ChIP-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="RNA-seq.html">RNA-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="scRNA-seq.html">scRNA-seq</a></li>
<li class="toctree-l1"><a class="reference internal" href="HiC.html">HiC</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">WGBS</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-requirements-and-outputs">Input requirements and outputs:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#configuration-file">Configuration file</a></li>
<li class="toctree-l2"><a class="reference internal" href="#structure-of-output-directory">Structure of output directory</a></li>
<li class="toctree-l2"><a class="reference internal" href="#usage-example">Usage example</a></li>
<li class="toctree-l2"><a class="reference internal" href="#argparse">Argparse</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Positional Arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#required arguments">required arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#general arguments">general arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#optional">optional</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../ChangeLog.html">Change Log</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">snakePipes</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
      <li>WGBS</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/content/workflows/WGBS.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="wgbs">
<span id="id1"></span><h1>WGBS<a class="headerlink" href="#wgbs" title="Permalink to this headline">¶</a></h1>
<div class="section" id="input-requirements-and-outputs">
<h2>Input requirements and outputs:<a class="headerlink" href="#input-requirements-and-outputs" title="Permalink to this headline">¶</a></h2>
<p>This pipeline requires paired-end reads fastq files and a bisulfite converted genome as inputs. Methylation values per CpG dinucleotide are extracted after read mapping.
Statistical analysis of differential methylation is performed if user provides a bed file with genomic intervals to aggregate methylation values over. De novo DMR calling, statistical evaluation and annotation with nearest gene are performed automatically. Various quality metrics are collected and a QC report is output.</p>
</div>
<div class="section" id="configuration-file">
<h2>Configuration file<a class="headerlink" href="#configuration-file" title="Permalink to this headline">¶</a></h2>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat snakemake_workflows/workflows/WGBS/defaults.yaml
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">################################################################################</span>
<span class="c1"># This file is the default configuration of the WGBS workflow!</span>
<span class="c1">#</span>
<span class="c1"># In order to adjust some parameters, please either use the wrapper script</span>
<span class="c1"># (eg. /path/to/snakemake_workflows/workflows/WGBS/WGBS)</span>
<span class="c1"># or save a copy of this file, modify necessary parameters and then provide</span>
<span class="c1"># this file to the wrapper or snakemake via &#39;--configfile&#39; option</span>
<span class="c1"># (see below how to call the snakefile directly)</span>
<span class="c1">#</span>
<span class="c1"># Own parameters will be loaded during snakefile executiuon as well and hence</span>
<span class="c1"># can be used in new/extended snakemake rules!</span>
<span class="c1">################################################################################</span>
<span class="c1">## General/Snakemake parameters, only used/set by wrapper or in Snakemake cmdl, but not in Snakefile</span>
<span class="n">wdir</span><span class="p">:</span>
<span class="n">configfile</span><span class="p">:</span>
<span class="n">cluster_configfile</span><span class="p">:</span>
<span class="n">local</span><span class="p">:</span> <span class="kc">False</span>
<span class="n">max_jobs</span><span class="p">:</span> <span class="mi">12</span>
<span class="n">nthreads</span><span class="p">:</span> <span class="mi">8</span>
<span class="n">snakemake_options</span><span class="p">:</span> <span class="s1">&#39;--use-conda --conda-prefix /data/processing/conda/envs&#39;</span>
<span class="n">tempdir</span><span class="p">:</span> <span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">extended</span><span class="o">/</span>
<span class="c1">## directory with fastq files</span>
<span class="n">readin</span><span class="p">:</span>
<span class="c1">## directory with fastqc files for auto quality trimming</span>
<span class="n">fqcin</span><span class="p">:</span>
<span class="c1">## Genome information</span>
<span class="n">genome</span><span class="p">:</span>
<span class="n">convrefpath</span><span class="p">:</span>
<span class="n">convRef</span><span class="p">:</span> <span class="kc">False</span>
<span class="c1">###list of bed files to process</span>
<span class="n">intList</span><span class="p">:</span> <span class="p">[]</span>
<span class="c1">###SNP black list (bed file)</span>
<span class="n">blackList</span><span class="p">:</span>
<span class="c1">###sample Info</span>
<span class="n">sampleInfo</span><span class="p">:</span>
<span class="c1">###inclusion bounds for methylation extraction</span>
<span class="n">mbias_ignore</span><span class="p">:</span> <span class="n">auto</span>
<span class="c1">## FASTQ file extension (default: &quot;.fastq.gz&quot;)</span>
<span class="n">ext</span><span class="p">:</span> <span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
<span class="c1">## paired-end read name extension (default: [&#39;_R1&#39;, &quot;_R2&quot;])</span>
<span class="n">reads</span><span class="p">:</span> <span class="p">[</span><span class="n">_R1</span><span class="p">,</span> <span class="n">_R2</span><span class="p">]</span>
<span class="c1">## Number of reads to downsample from each FASTQ file</span>
<span class="n">downsample</span><span class="p">:</span>
<span class="c1">## Options for trimming</span>
<span class="n">trimReads</span><span class="p">:</span> <span class="n">auto</span>
<span class="n">nextera</span><span class="p">:</span> <span class="kc">False</span>
<span class="n">trimThreshold</span><span class="p">:</span> <span class="mi">10</span>
<span class="n">trimOtherArgs</span><span class="p">:</span>
<span class="n">verbose</span><span class="p">:</span> <span class="kc">False</span>
<span class="c1">################################################################################</span>
<span class="c1"># Call snakemake directly, i.e. without using the wrapper script:</span>
<span class="c1">#</span>
<span class="c1"># Please save a copy of this config yaml file and provide an adjusted config</span>
<span class="c1"># via &#39;--configfile&#39; parameter!</span>
<span class="c1"># example call:</span>
<span class="c1">#</span>
<span class="c1"># snakemake --snakefile /path/to/snakemake_workflows/workflows/WGBS/Snakefile</span>
<span class="c1">#           --configfile /path/to/snakemake_workflows/workflows/WGBS/defaults.yaml</span>
<span class="c1">#           --directory /path/to/outputdir</span>
<span class="c1">#           --cores 32</span>
<span class="c1">################################################################################</span>
</pre></div>
</div>
</div>
<div class="section" id="structure-of-output-directory">
<h2>Structure of output directory<a class="headerlink" href="#structure-of-output-directory" title="Permalink to this headline">¶</a></h2>
<p>The WGBS pipeline will generate additional output as follows:</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span>$ tree -d -L 2 output-dir/
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>output_dir
|-- FASTQ
|-- FASTQ_Cutadapt
|   `-- logs
|-- FASTQ_downsampled
|   `-- logs
|-- FastQC_Cutadapt
|   `-- logs
|-- FastQC_In
...
|   `-- logs
|-- QC_metrics
|   `-- logs
|-- aggregate_stats_limma
|   `-- logs
|-- aux_files
|   `-- logs
|-- bams
|   `-- logs
|-- cluster_logs
|-- methXT
|   `-- logs
|-- metilene_out
|   `-- logs
`-- singleCpG_stats_limma
    `-- logs
</pre></div>
</div>
<p>Aggregate stats will be calculated if user provides at least one bed file with genomic intervals of interest. Differential methylation analysis or DMR detection will only be run if user provides a sample sheet.</p>
</div>
<div class="section" id="usage-example">
<h2>Usage example<a class="headerlink" href="#usage-example" title="Permalink to this headline">¶</a></h2>
<p>python3 snakemake_workflows/workflows/WGBS/WGBS --indir /data/manke/group/snakepipes_test_files/WGBS/reads --wdir /data/processing3/WGBS_snakepipe_example_OUT_logging_auto --intList /data/manke/group/snakepipes_test_files/WGBS/danRer10.cpgIsland.ext.sorted.chr25.noCHR.bed --sampleInfo /data/manke/group/snakepipes_test_files/WGBS/example_sampleSheet.csv GRCz10</p>
</div>
<div class="section" id="argparse">
<h2>Argparse<a class="headerlink" href="#argparse" title="Permalink to this headline">¶</a></h2>
<p><p>WGBS workflow v0.7.1 - MPI-IE workflow for WGBS analysis</p>
<dl class="docutils">
<dt>usage example:</dt>
<dd>WGBS -ri read_input_dir -w output-dir mm10</dd>
</dl>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">WGBS</span> <span class="o">-</span><span class="n">ri</span> <span class="n">INDIR</span> <span class="o">-</span><span class="n">w</span> <span class="n">WDIR</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">CONFIGFILE</span><span class="p">]</span>
            <span class="p">[</span><span class="o">--</span><span class="n">cluster_configfile</span> <span class="n">CLUSTER_CONFIGFILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">j</span> <span class="n">INT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">nt</span> <span class="n">INT</span><span class="p">]</span>
            <span class="p">[</span><span class="o">--</span><span class="n">local</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">snakemake_options</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">tempdir</span> <span class="n">TEMPDIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">DAG</span><span class="p">]</span>
            <span class="p">[</span><span class="o">--</span><span class="n">downsample</span> <span class="n">INT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">trimReads</span> <span class="p">{</span><span class="n">auto</span><span class="p">,</span><span class="n">user</span><span class="p">,</span><span class="kc">None</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">fqcin</span> <span class="n">FQCIN</span><span class="p">]</span>
            <span class="p">[</span><span class="o">--</span><span class="n">nextera</span> <span class="n">NEXTERA</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">trimThreshold</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">trimOtherArgs</span> <span class="n">STR</span><span class="p">]</span>
            <span class="p">[</span><span class="o">--</span><span class="n">cref</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">convRef</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">intList</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">blackList</span> <span class="n">STR</span><span class="p">]</span>
            <span class="p">[</span><span class="o">--</span><span class="n">sampleInfo</span> <span class="n">STR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">mbias</span> <span class="n">STR</span><span class="p">]</span>
            <span class="n">GENOME</span>
</pre></div>
</div>
<div class="section" id="Positional Arguments">
<h3>Positional Arguments<a class="headerlink" href="#Positional Arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>GENOME</kbd></td>
<td>Genome acronym of target organism (supported: 'dm3', 'dm6', 'hs37d5', 'mm9', 'mm10', 'SchizoSPombe_ASM294v2')</td></tr>
</tbody>
</table>
</div>
<div class="section" id="required arguments">
<h3>required arguments<a class="headerlink" href="#required arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-ri, --indir</kbd></td>
<td>input directory containing the paired-end FASTQ files</td></tr>
<tr><td class="option-group">
<kbd>-w, --wdir</kbd></td>
<td>working directory</td></tr>
</tbody>
</table>
</div>
<div class="section" id="general arguments">
<h3>general arguments<a class="headerlink" href="#general arguments" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>-v, --verbose</kbd></td>
<td>verbose output (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-c, --configfile</kbd></td>
</tr>
<tr><td>&#160;</td><td>configuration file: config.yaml. In absence of a config file, the default options from the workflows would be selected from /workflows/&lt;workflow_name&gt;/defaults.yaml (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--cluster_configfile</kbd></td>
</tr>
<tr><td>&#160;</td><td>configuration file for cluster usage. In absence, the default options from shared/cluster.yaml and workflows/[workflow]/cluster.yaml would be selected (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>-j, --jobs</kbd></td>
<td>maximum number of concurrently submitted Slurm jobs / cores if workflow is run locally (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>-nt, --nthreads</kbd></td>
</tr>
<tr><td>&#160;</td><td>number of threads to use for multi-threaded tasks</td></tr>
<tr><td class="option-group">
<kbd>--local</kbd></td>
<td><p class="first">run workflow locally; default: jobs are submitted to Slurm queue (default: 'False')</p>
<p class="last">Default: False</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--snakemake_options</kbd></td>
</tr>
<tr><td>&#160;</td><td>Snakemake options to be passed directly to snakemake, e.g. use --snakemake_options='--dryrun --rerun-incomplete --unlock --forceall'. (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--tempdir</kbd></td>
<td>used prefix path for temporary directory created via mktemp. Created temp dir gets exported as $TMPDIR and is removed at the end of this wrapper! (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--DAG</kbd></td>
<td><p class="first">If specified, a file called pipeline.pdf is produced in the output directory that shows the rules used and their relationship to each other.</p>
<p class="last">Default: False</p>
</td></tr>
</tbody>
</table>
</div>
<div class="section" id="optional">
<h3>optional<a class="headerlink" href="#optional" title="Permalink to this headline">¶</a></h3>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>--downsample</kbd></td>
<td>downsample the given number of reads randomly from of each FASTQ file (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--trimReads</kbd></td>
<td><p class="first">Possible choices: auto, user, None</p>
<p class="last">Activate fastq read trimming. If activated, Illumina adaptors are trimmed by default. Additional parameters can be specified under --trimOtherArgs (default: 'None')</p>
</td></tr>
<tr><td class="option-group">
<kbd>--fqcin</kbd></td>
<td>Folder with fastqc.zip files to automatically detect hard trimming threshold.</td></tr>
<tr><td class="option-group">
<kbd>--nextera</kbd></td>
<td>Trim nextera adapters rather than TruSeq adapters (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--trimThreshold</kbd></td>
</tr>
<tr><td>&#160;</td><td>Trimming threshold for cutadapt. (default: 'None')</td></tr>
<tr><td class="option-group" colspan="2">
<kbd>--trimOtherArgs</kbd></td>
</tr>
<tr><td>&#160;</td><td>Additional arguments passed to cutadapt. (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--cref</kbd></td>
<td>Path to converted reference genome. (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--convRef</kbd></td>
<td>Convert reference genome. (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--intList</kbd></td>
<td>Bed file(s) with target intervals to run analysis on. Use multiple times to pass multiple bed files. (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--blackList</kbd></td>
<td>Bed file(s) with SNP positions to mask for methylation calling. (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--sampleInfo</kbd></td>
<td>Text file with sample information to use for statistical analysis. (default: 'None')</td></tr>
<tr><td class="option-group">
<kbd>--mbias</kbd></td>
<td>Number of nucleotides with mbias to ignore during methylation extraction. (default: 'None')</td></tr>
</tbody>
</table>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference external" href="https://github.com/maxplanck-ie/snakemake_workflows">code &#64; github</a>.</td>
</tr>
</tbody>
</table>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../../ChangeLog.html" class="btn btn-neutral float-right" title="Change Log" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="HiC.html" class="btn btn-neutral" title="HiC" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, MPI-IE.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.6.2',
            LANGUAGE:'python',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>